### ISSUE TITLE
How does the audio emulation work?

### ISSUE THREAD

**Issue by cbmeeks** [visrealm/pico-56#12]
I realize you forked the emu2149 library and added more code for the Pico.

It seems to be pretty straightforward to get the audio piece to work (using AY-3-8910 mode).  However, nothing seems to work on my setup.

I suspect that you have to call the `audioUpdate()` very quickly?  I believe you are calling it at the end of each horizontal scanline.
Unfortunately, that won't work for my project as I need a more "general purpose" implementation (called from a 6502).

At the very least, what I am doing is something like:

    // setup code
    audioInit(3579545 / 2, 44100);
    audioWritePsg0(7, 0b00111110);
    audioWritePsg0(8, 0b00001111);

    // main loop
     audioWritePsg0(0, 223);
    audioWritePsg0(1, 1);
    audioUpdate();

I really just got those register values from a YouTube video where a guy controls a real AY with an Arduino (which I did years ago).

Do you know what I am missing?

Thanks for any input.

**Comment by visrealm** [comment]
audioUpdate() needs to be called at a regular frequency matching the sampleRate. You can call it more frequently,  but you'd need to buffer the psg->ch_out values after each call and apply them to your pwm outputs at a regular frequency. 
If you can't use the hsync, you'll need to either use a timer or poll the current time and update the pwm values when the next sample is due.

**Comment by visrealm** [comment]
Also, audioWritePsg0() doesn't work like your example expects. It's set up like the PSG is on the bus. To write a register value requires two calls to audioWritePsg0() just like the real device. First with LSB of 0 to set the register number, then with an LSB of 1 to set the value.

So,

audioWritePsg0(7, 0b00111110);

Should be

audioWritePsg0(0, 7);
audioWritePsg0(1, 0b00111110);

Seems odd, but that's what my HBC-56 has to do to write to a PSG (except 0 and 1 are replaced with the physical address of the PSG). The underlying library has a function  to write a value to a register in one call though.

**Comment by cbmeeks** [comment]
OK, I think that makes sense.  Years ago I controlled a real AY-3-8910 with an Arduino and the difference here is that it had a real clock.  Since this is in software, we basically have to emulate that.

I will give that a shot.

Thanks again!

**Comment by cbmeeks** [comment]
For those that might be interested, this seems to work!

At a sample rate of 44,100 Hz, I used the following:

1000 (ms) / 44100 = 0.022675737
0.022675737 * 1000 (uS) = 22.675736961 uS

I rounded up and just used a 23uS repeating timer:

    bool repeating_timer_callback(struct repeating_timer *t) {
        audioUpdate();
        return true;
    }


    add_repeating_timer_us(23, repeating_timer_callback, NULL, &timer);

    while (1) {
        
        // plays a middle A (440)
        audioWritePsg0(0, 0);   // register 0
        audioWritePsg0(1, 28);   // register 0 value

        audioWritePsg0(0, 1);   // register 1
        audioWritePsg0(1, 1);   // register 1 value

        sleep_ms(500);
    }

**Comment by cbmeeks** [comment]
> audioUpdate() needs to be called at a regular frequency matching the sampleRate. You can call it more frequently, but you'd need to buffer the psg->ch_out values after each call and apply them to your pwm outputs at a regular frequency. If you can't use the hsync, you'll need to either use a timer or poll the current time and update the pwm values when the next sample is due.

I'm playing some VGM music but something is off.  It only works if I stream the registers/data at the same rate of the `audioUpdate`.  Is that correct?  Basically, does the audioUpdate the only thing that needs to be called at the same frequency or does sending commands also have to be at that rate?

Thanks.

UPDATE.

Maybe I answered my own question.  Changing the frequency from 1/2 NTSC to 4MHz sounds more accurate.  But still not perfect.

At least I can tweak these numbers.  I wonder, however, how accurate the emu2149 is.  When you listen to VGM's from https://vgmrips.net/packs/pack/vampire-killer-msx2 you can tell there is a difference.

I will post my repo soon if anyone is interested.

Thanks!

**Comment by visrealm** [comment]
Note the MSX2 had the YM2149, so it might be that it hasn't been initialized as a YM2149, but rather an AY-3-8910. You might have changed this already, but just in case...

See this from the README:
```

/* Typical YM2149 configuration */
PSG *psg = PSG_new(3579545, 44100);
PSG_setClockDivider(psg, 1);
PSG_setVolumeMode(psg, 1); // YM style
PSG_reset();

/* Typical AY-3-8910 configuration */
PSG *psg = PSG_new(3579545/2, 44100);
PSG_setVolumeMode(psg, 2); // AY style
PSG_reset();
```

**Comment by cbmeeks** [comment]
Ah, I didn't see the `PSG_setClockDivider` method.  Also, it's not exposed in the header file of emu2149.h that I saw but I corrected it.

I did originally try setting the volume mode to the YM but didn't notice a difference.  

I have a project that uses the library and plays a VGM song if you want to listen for yourself.

Either way, I will continue messing around with this library as I really like having it in my project.  :-)

https://github.com/cbmeeks/PicoSnd
